evm/src/cpu/kernel/asm/bignum/modmul.asm

```diff
@@ -0,0 +1,192 @@
+    PUSH 0
+    // stack: i=0, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+modmul_quotient_loop:
+    // stack: i, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    PROVER_INPUT(bignum_modmul::quotient)
+    // stack: PI, i, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    DUP8
+    // stack: scratch_1, PI, i, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    DUP3
+    // stack: i, scratch_1, PI, i, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    ADD
+    // stack: scratch_1[i], PI, i, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    %mstore_kernel_general
+    // stack: i, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    %increment
+    // stack: i+1, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    DUP2
+    DUP2
+    // stack: i+1, length, i+1, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    %neq
+    // stack: i+1!=length, i+1, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    %jumpi(modmul_quotient_loop)
+modmul_quotient_end:
+    // stack: i, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    POP
+    // stack: length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+
+    // Verification step 1: calculate x + k * m.
+
+    // Store k * m in scratch_2.
+    PUSH modmul_return_1
+    // stack: modmul_return_1, length, a_start_loc, b_start_loc, m_start_loc, output_loc, scratch_1, scratch_2, scratch_3, retdest
+    %stack (return, len, a, b, m, out, s1, s2) -> (len, s1, m, s2, return, len, a, b, out, s2)
+    // stack: length, scratch_1, m_start_loc, scratch_2, modmul_return_1, length, a_start_loc, b_start_loc, output_loc, scratch_2, scratch_3, retdest
+    %jump(mul_bignum)
+modmul_return_1:
+    // stack: length, a_start_loc, b_start_loc, output_loc, scratch_2, scratch_3, retdest
+
+    // Add x into k * m (in scratch_2).
+    PUSH modmul_return_2
+    // stack: modmul_return_2, length, a_start_loc, b_start_loc, output_loc, scratch_2, scratch_3, retdest
+    %stack (return, len, a, b, out, s2) -> (len, s2, out, return, len, a, b, s2)
+    // stack: length, scratch_2, output_loc, modmul_return_2, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    %jump(add_bignum)
+modmul_return_2:
+    // stack: carry, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    ISZERO
+    %jumpi(no_carry)
```

Comment:
Wait, can you ever have a carry? `scratch_2` has `2 * 128 * length` bits, so it should fit all values `<= 2^(2 * 128 * length) - 1`. Now, `k`, `m`, and `x` are all `<= 2^(128 * length) - 1`, so  `k * m <= 2^(2 * 128 * length) - 2 * 2^(128 * length + 1) + 1` and `x + k * m <= 2^(2 * 128 * length) - 2^(128 * length)`, which is `<= 2^(128 * length) - 1`. Am I missing something?

Reply:
In any case, if there's a carry, then shouldn't we just PANIC? Because `a * b` definitely fits in `2 * 128 * length` bits, so if `x + k * m` doesn't, then they can't be equal.

Reply:
Good point!!